controllers/openstackcontrolplane_controller.go

```diff
@@ -695,20 +591,454 @@ func (r *OpenStackControlPlaneReconciler) createOrUpdateOpenStackMACAddress(inst
+				cond.Reason = ospdirectorv1beta1.ConditionReason(ospdirectorv1beta1.CommonCondReasonConfigMapMissing)
+				cond.Type = ospdirectorv1beta1.ConditionType(ospdirectorv1beta1.CommonCondTypeWaiting)
+
+				common.LogForObject(r, cond.Message, instance)
+
+				return ctrl.Result{RequeueAfter: time.Duration(timeout) * time.Second}, nil
+			}
+			// Error reading the object - requeue the request.
+			cond.Message = fmt.Sprintf("Error reading config map object: %s", configMapName)
+			cond.Reason = ospdirectorv1beta1.ConditionReason(ospdirectorv1beta1.CommonCondReasonConfigMapError)
+			cond.Type = ospdirectorv1beta1.ConditionType(ospdirectorv1beta1.CommonCondTypeError)
+			err = common.WrapErrorForObject(cond.Message, instance, err)
+
+			return ctrl.Result{}, err
+		}
+		common.LogForObject(
+			r,
+			fmt.Sprintf("Secret %s exists", configMapName),
+			instance,
+		)
+	}
+	return ctrl.Result{}, nil
+}
+
+//
+// create hostnames for the requested number of systems
+//
+func (r *OpenStackControlPlaneReconciler) createNewHostnames(
+	instance *ospdirectorv1beta1.OpenStackControlPlane,
+	cond *ospdirectorv1beta1.Condition,
+	baseName string,
+	newCount int,
+	vip bool,
+) ([]string, error) {
+	newVMs := []string{}
+
+	//
+	//   create hostnames for the newCount
+	//
+	currentNetStatus := instance.Status.DeepCopy().VIPStatus
+	for i := 0; i < newCount; i++ {
+		hostnameDetails := common.Hostname{
+			Basename: baseName,
+			VIP:      vip,
+		}
+
+		err := common.CreateOrGetHostname(instance, &hostnameDetails)
+		if err != nil {
+			cond.Message = fmt.Sprintf("error creating new hostname %v", hostnameDetails)
+			cond.Reason = ospdirectorv1beta1.ConditionReason(ospdirectorv1beta1.CommonCondReasonNewHostnameError)
+			cond.Type = ospdirectorv1beta1.ConditionType(ospdirectorv1beta1.CommonCondTypeError)
+			err = common.WrapErrorForObject(cond.Message, instance, err)
+
+			return newVMs, err
+		}
+
+		if hostnameDetails.Hostname != "" {
+			if _, ok := instance.Status.VIPStatus[hostnameDetails.Hostname]; !ok {
+				instance.Status.VIPStatus[hostnameDetails.Hostname] = ospdirectorv1beta1.HostStatus{
+					Hostname:             hostnameDetails.Hostname,
+					HostRef:              hostnameDetails.HostRef,
+					AnnotatedForDeletion: false,
+					IPAddresses:          map[string]string{},
+				}
+				newVMs = append(newVMs, hostnameDetails.Hostname)
+			}
+
+			common.LogForObject(
+				r,
+				fmt.Sprintf("%s hostname created: %s", instance.Kind, hostnameDetails.Hostname),
+				instance,
+			)
+		}
+	}
+
+	if !reflect.DeepEqual(currentNetStatus, instance.Status.VIPStatus) {
+		common.LogForObject(
+			r,
+			fmt.Sprintf("Updating CR status with new hostname information, %d new - %s",
+				len(newVMs),
+				diff.ObjectReflectDiff(currentNetStatus, instance.Status.VIPStatus),
+			),
+			instance,
+		)
+
+		err := r.Client.Status().Update(context.TODO(), instance)
```

Comment:
Would the defer in the main reconcile loop not try to update the status anyhow, since it would compare `Status` to the old deep copy and see the difference(s)?

Reply:
yes it will when we return. since the hostname is some important information to create the IPs for I thought it might make sense to update it right after created them and fail/reconcile early if it fail for whatever reasone.

Reply:
Ack, makes sense